### PR TITLE
test: add showcase test for api-version

### PR DESCRIPTION
**Pull request by zhumin8**
This pr updates gapic-showcase version to 0.35.0 and adds showcase tests to verify behavior of api version headers being emitted (changes in #2630 and #2671).

### CODE REVIEW

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                    metadata = trailers;
+                  }
+                  super.onClose(status, trailers);
+                }
+              };
+
+          super.start(wrappedListener, headers);
+        }
+      };
+    }
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:63]: Not something that has to be done in this PR, but since using client interceptors is something that multiple tests have implemented, I think we can try and migrate this to TestClientInitializer if possible (I don't know if there are an limitations with that, but just something to consider).

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.showcase.v1beta1.*;
+import com.google.showcase.v1beta1.it.util.TestClientInitializer;
+import io.grpc.*;
+import java.util.ArrayList;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ITApiVersionHeaders {
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:30]: Can you add a quick comment, like a TODO, above this that we can use echo to echo back the once this gapic-showcase PR makes it in? https://github.com/googleapis/gapic-showcase/pull/1456

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  private static final String SPLIT_TOKEN = "&";
+  private static final String API_VERSION_HEADER_STRING = "x-goog-api-version";
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:32]: nit: Personal opinion is that we can consider making the constant public: https://github.com/googleapis/sdk-platform-java/blob/43e62b998958abd4ca96b83800380c9696ff693f/gax-java/gax/src/main/java/com/google/api/gax/rpc/ApiClientHeaderProvider.java#L45 so that we don't redefine it multiple spots. 

We tried package-private, but showcase needs it as well.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    httpJsonClient.close();
+    grpcComplianceClient.close();
+    httpJsonComplianceClient.close();
+  }
+
+  @Test
+  public void testGrpc_matchesApiVersion() {
+    grpcClient.echo(EchoRequest.newBuilder().build());
+    String headerValue = grpcInterceptor.metadata.get(API_VERSION_HEADER_KEY);
+    assertThat(headerValue).isEqualTo("v1_20240408");
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:180]: nit: Perhaps a constant value for here and the one below? i.e. something like `DEFAULT_SHOWCASE_API_VERSION`?

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 1 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:63]: mmh, sounds a good idea. These interceptors are slightly different, but I guess we can create an interceptor that does it all.
To keep current pr from being cluttered, I would file a separate issue to look into and refactor these.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 2 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:30]: Will do. I'll probably also file an issue to track, so we don't forget about it.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    grpcComplianceInterceptor = new GrpcCapturingClientInterceptor();
+    grpcComplianceClient =
+        TestClientInitializer.createGrpcComplianceClient(
+            ImmutableList.of(grpcComplianceInterceptor));
+
+    // Create HttpJson ComplianceClient and Interceptor
+    httpJsonComplianceInterceptor = new HttpJsonCapturingClientInterceptor();
+    httpJsonComplianceClient =
+        TestClientInitializer.createHttpJsonComplianceClient(
+            ImmutableList.of(httpJsonComplianceInterceptor));
```
[the remark this answers is not in the document]
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:165]: In current setup, we _could_ override with `.setHeaderProvider()` when creating the test client. I am reluctant to do so though. 
- I feel like showcase integration testing should test expected behavior of un-modified service behaviors E2E.
- As we discussed in doc comment last time, I am inclined to not allow user override this "x-goog-api-version" header with client libraries. (Will raise a separate pr) In this case, "unsets the showcase apiVersion value" should not work.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 3 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:32]: I tend to agree. But I was not 100% on if this a breaking change to gax. Even though we are exposing a new public field, this is a final field and no breaking to existing usage, so I suppose it's okay?

Comment (lqiu96) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java]: It was package-private so we can change it since no one else should have access to it.

It is now public field, but I think that's fine since we're doing it so that this constant isn't re-defined in multiple spots.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  @After
+  public void destroyClient() {
+    grpcClient.close();
+    httpJsonClient.close();
+    grpcComplianceClient.close();
+    httpJsonComplianceClient.close();
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:179]: Can you also add `awaitTermination()` here after the `.close()` calls?

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lqiu96) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:171]: > I feel like showcase integration testing should test expected behavior of un-modified service behaviors E2E.

IMO, it shouldn't just be limited to un-modified service behaviors. We should try to account for customizations where we can (i.e. If new features are added or showcase has new functionality). We're not in state where every modification/ feature is covered, but I think we should strive to get as close as possible. There are going to be certain spots we miss and new edge cases that are discovered, but we should try to backfill/ add them in as we discover them.

> I am inclined to not allow user override this "x-goog-api-version" header with client libraries.

Makes sense. Let's get some insight from the core team to see if this is prioritized (and if not, we can always prioritize this for Java clients ourselves). The public setter method is pubic as of now and removing will be challenging/ impossible. We can explore some options to see how we want to proceed.

Being that removing may be challenging/impossible, I think it may make sense to include both Echo and Compliance clients in then since we test for the behavior of add custom headers to both clients that that have an auto generated ApiVersion vs clients that don't have this generated.

i.e. I believe Echo with custom headers set should have a conflict error message and Compliance with custom headers should not have a conflict error message.

[the remark this answers is not in the document]
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:171]: I agree with most of these, but wanted to clarify a few:
- This [ITApiVersionHeaders.java](https://github.com/googleapis/sdk-platform-java/pull/2737/files/219c8fdb150a33a9f2e44c442f4e91a586000ead#diff-0c74baec42410162926db3ec4be8e8f2c04fd9d6cf214ef5bbf407e7e350f0b5) tests for behavior when client has an auto generated ApiVersion (Echo) and gets correctly send via header on requests, and when client don't have this generated (Compliance) then no such header sent. User header are not covered.
- To test behavior of adding user header to both clients also makes sense, I will add based on current behavior. This may change in the future. In fact, I have a todo item in https://github.com/googleapis/sdk-platform-java/pull/2745 for it if we decide to go with this approach.
- "The public setter method" >> I assume you are referring to [this setter](https://github.com/googleapis/sdk-platform-java/blob/c9d8ded1a8af3d7cbb9a647cb802a71412c8d7d7/gax-java/gax/src/main/java/com/google/api/gax/rpc/ApiClientHeaderProvider.java#L218)? If so, this is only one HeaderProvider the client library code uses, users could use other HeaderProvider. But to guard users to not override  "x-goog-api-version" header with `setHeaderProvider()`, I am thinking adding to conflict resolution as in [pr2745](https://github.com/googleapis/sdk-platform-java/pull/2745) 
- "Let's get some insight from the core team if this is prioritized " >> Other than comment on proposal doc, I've started a thread on the aip [here](https://github.com/aip-dev/google.aip.dev/pull/1331/files#r1591764838) 

To summarize, I am keeping both clients in this test to cover scenario of service opted-in for api-version and not. Also adding tests for setting custom api version headers for these clients, these behavior may change via a separate pr.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 6 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:185]: IIUC, `.close()` is same as `.shutdown()` ([code](https://github.com/googleapis/sdk-platform-java/blob/8381dc58c3bc0d5821c9c22bb7af49b3ad9e544e/gax-java/gax/src/main/java/com/google/api/gax/core/BackgroundResourceAggregation.java#L93)). We should `.shutdown()` a client or `.awaitTermination()`, not both, and the main difference is that `.awaitTermination()` has timeout ([comment here](https://github.com/googleapis/sdk-platform-java/blob/8381dc58c3bc0d5821c9c22bb7af49b3ad9e544e/gax-java/gax/src/main/java/com/google/api/gax/core/BackgroundResource.java#L70-L74)).
However, if we want to prevent a test hanging, this may not work, as these are called within `@After`  and I believe if a test hangs and does not finish, the `@After` method will not be executed. An alternative is to add `@Test(timeout = ..)` to tests. But I wonder if you have a specific concern about these tests? Or do you prefer to implement timeouts for all integration tests as precautious?

Comment (lqiu96) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:185]: We typically use `close()` in conjunction with `awaitTermination()`, with `close()` being called first. You're right `close()` just calls `shutdown()` and all `shutdown()` does to prevent *new* tasks from being scheduled, but still allows currently in-flight tasks to continue. We simply use `awaitTermination()` to allow the current in-flight tasks to finish (it blocks until they are done or until the timeout has been reached).

We had previously seen some grpc-java warnings about managed channels still being allocated without trying to wait to the tasks to terminated. i.e something like:
```
io.grpc.internal.ManagedChannelImpl$ManagedChannelReference cleanQueue
SEVERE: *~*~*~ Channel io.grpc.internal.ManagedChannelImpl-2096 for target 127.0.0.1:8771 was not shutdown properly!!! ~*~*~*
    Make sure to call shutdown()/shutdownNow() and awaitTermination().
java.lang.RuntimeException: ManagedChannel allocation site
```

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lqiu96) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:177]: > This [ITApiVersionHeaders.java](https://github.com/googleapis/sdk-platform-java/pull/2737/files/219c8fdb150a33a9f2e44c442f4e91a586000ead#diff-0c74baec42410162926db3ec4be8e8f2c04fd9d6cf214ef5bbf407e7e350f0b5) tests for behavior when client has an auto generated ApiVersion (Echo) and gets correctly send via header on requests, and when client don't have this generated (Compliance) then no such header sent. User header are not covered.

Yep, I think that's good! The test cases do look good. I don't mean that we should try to cover modifying *every* user header, but rather the relevant one to the test (api-version header).

> In fact, I have a todo item in https://github.com/googleapis/sdk-platform-java/pull/2745

Let's have an internal discussion. I do prefer this behavior but I believe this does technically put us at a "behavior breaking change" which is something that we try not to do (if possible). I think this may be a case where we can argue for it given that it's a relatively new feature.

> Also adding tests for setting custom api version headers for these clients, these behavior may change via a separate pr.

I'll do another round of review on this PR. No worries about adding modifications to the header in this PR if you haven't already. I believe it's already a pretty niche case and it can be added and/or tested based on what happens to #2745

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+  @Test
+  public void testHttpJson_noApiVersion() {
+    RepeatRequest request =
+        RepeatRequest.newBuilder().setInfo(ComplianceData.newBuilder().setFString("test")).build();
+    httpJsonComplianceClient.repeatDataSimplePath(request);
+    assertThat(API_VERSION_HEADER_KEY)
+        .isNotIn(httpJsonComplianceInterceptor.metadata.getHeaders().keySet());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:221]: Can you use `assertThrows()` instead of `@Test(expected = ...)`? Something like: https://github.com/googleapis/sdk-platform-java/blob/3611e8f8982070f6b1aa14b6272677608aa5f6fb/showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITUnaryCallable.java#L70-L71

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            .getStubSettings()
+            .toBuilder()
+            .setHeaderProvider(
+                FixedHeaderProvider.create(
+                    ApiClientHeaderProvider.API_VERSION_HEADER_KEY, CUSTOM_API_VERSION))
+            .build();
+
+    try (EchoClient echo =
+        EchoClient.create(EchoSettings.create((EchoStubSettings) stubSettings))) {
+      assertThat(true).isFalse();
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:235]: What is this line doing?

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    }
+  }
```
Comment (lqiu96) [inline showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:255]: Same as the two comments above.

**Comment by lqiu96** [comment]
LGTM. Just a few nits/ questions, but I'll approve once those are addressed.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 8 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:235]: Oh, nvm. 
I was not aware the `assertThrows()` as you mentioned in above comment is available in Junit 4 too, and was trying to be more precise on where exception is thrown. Switching to `assertThrows()` approach.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 9 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:255]: same above.

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 6 above)
Comment (lqiu96) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:189]: Bumping this.

**Review by lqiu96 — APPROVED** [review]
LGTM. Small addition for `@After` if you could resolve

File: showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java (hunk 6 above)
Comment (zhumin8) [reply showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITApiVersionHeaders.java:191]: Thanks for explaining. Added.
